### PR TITLE
docs: add migration guide

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR finishes https://github.com/asyncapi/parser-js/pull/611

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,7 @@
 node_modules
 .vscode
 .DS_Store
-/docs
+/docs/api.md
```
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply .gitignore:4]: I guess not any more no 😄

Comment (jonaslagoni) [inline .gitignore:4]:
```suggestion 
```

Comment (jonaslagoni) [inline .gitignore:4]:
```suggestion
```

Comment (jonaslagoni) [inline .gitignore:4]:
```suggestion
```

File: docs/migrations/v1-to-v2.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+- The way to pass custom `$ref` resolvers has changed. [Read more](#custom-reference-resolvers).
+- Stringify and unstringify the parsed document is now a separate functionality. [Read more](#standalone-stringify-and-unstringify-functionality).
+- If there is a need to use an old version of the API, the Parser provides the `convertToOldAPI` functionality. [Read more](#convert-new-api-to-the-old-one).
+
+### New Intent API
+
+The old Document API based 1:1 on AsyncAPI Specification has been rewritten to support the new so-called [Intent API](https://github.com/asyncapi/parser-api), which makes it easier to operate on the document and retrieve collections of needed objects from the document. Due to the fact that the new API does not stick firmly to AsyncAPI Specification, it allows for better integration with future versions of AsyncAPI. User does not need to know the new API against the new version of AsyncAPI - all methods will remain the same with the inclusion of new/deprecated fields. Also new API has been enhanced with some additional functionality like retrieving the JSON Path of the object using the `.jsonPath()` method etc.
```
Comment (smoya) [inline docs/migrations/v1-to-v2.md:19]: nitpicks 
```suggestion
The old Document API based 1:1 on AsyncAPI Specification has been rewritten to support the new so-called [Parser API](https://github.com/asyncapi/parser-api), which makes it easier to operate on the document and retrieve collections of needed objects from it. Due to the fact that the new API does not stick firmly to the AsyncAPI Specification file structure, it allows for better integration with future versions of AsyncAPI. The user does not need to know the new API against the new version of AsyncAPI - all methods will remain the same with the inclusion of new/deprecated fields. Also new API has been enhanced with some additional functionality like retrieving the JSON Path of the object using the `.jsonPath()` method etc.
```

File: docs/migrations/v1-to-v2.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+```ts
+import { Parser } from '@asyncapi/parser';
+
+const parser = new Parser();
+
+const { document } = await parser.parse(`...AsyncAPI document`, ...?options);
+```
+
+### Powered by [Spectral](https://github.com/stoplightio/spectral)
```
Comment (smoya) [inline docs/migrations/v1-to-v2.md:37]:
```suggestion
### Powered by Spectral
```

You already added the link after.

File: docs/migrations/v1-to-v2.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+[Spectral](https://github.com/stoplightio/spectral) is an open-source API linting tool by [Stoplight](https://stoplight.io/) that allows users to create custom style guides to enforce validations on API design automatically. This allows for better validation and showing all errors in one validation process - the old parser had a huge problem with this. In addition, there is the ability to show good practices though like defining `operationId` for each operation, or enforcing things that are optional but required for a project or company (API governance). Together with [Stoplight](https://stoplight.io/), we have an official partnership and will be improving integration as well as [Spectral](https://github.com/stoplightio/spectral) itself for better quality validation of AsyncAPI documents.
+
+### `.parse()` method
```
Comment (smoya) [inline docs/migrations/v1-to-v2.md:41]: I would unify all the bullets that are explaining methods (such as this parse() one) under a shared section. I.e. `## Important API changes` or similar.

File: docs/migrations/v1-to-v2.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+- It validates AsyncAPI documents using [Spectral](https://github.com/stoplightio/spectral). This enables better validation and introduces linting capabilities. [Read more](#spectral).
+- Introduces validation without parsing via the `.validate()` method. [Read more](#validate-method).
+- The `.parse()` method now also returns diagnostics (errors, warnings, and bits of advice). [Read more](#parse-method).
+- The `parseFromUrl()` method has been deprecated in favor of the `fromURL` standalone function. [Read more](#parseFromUrl-is-now-deprecated).
+- Custom parsers now have a `validate()` method. [Read more](#new-custom-schema-parser-interface).
+- The way to pass custom `$ref` resolvers has changed. [Read more](#custom-reference-resolvers).
+- Stringify and unstringify the parsed document is now a separate functionality. [Read more](#standalone-stringify-and-unstringify-functionality).
+- If there is a need to use an old version of the API, the Parser provides the `convertToOldAPI` functionality. [Read more](#convert-new-api-to-the-old-one).
+
+### New Intent API
```
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply docs/migrations/v1-to-v2.md:17]: I dont understand what you are suggesting 😅

**Comment by jonaslagoni** [comment]
Think I solved both your comments (including https://github.com/asyncapi/parser-js/pull/789#discussion_r1235194095) @smoya let me know if not 😄

**Review by smoya — DISMISSED** [review]
LGTM! 🚀🌔

**Comment by jonaslagoni** [comment]
@smoya can you disable sonarcloud required check 🙂

**Comment by jonaslagoni** [comment]
Or merge it manually.

**Comment by smoya** [comment]
> @smoya can you disable sonarcloud required check 🙂

I think I can't. The issue is with duplicated lines. TBH I don't really understand why the analysis is not limited to the changes added in the PR...

I can't manually merge it either 🤷 
Google Chrome_OltFSYoB

**Comment by jonaslagoni** [comment]
Cc admin @derberg

**Comment by derberg** [comment]
I think it is all about tests and it is completely fine if you just add `sonar.exclusions=test/**/*` to https://github.com/asyncapi/parser-js/blob/master/.sonarcloud.properties

**Review by smoya — APPROVED** [review]

**Comment by jonaslagoni** [comment]
Thanks @smoya 👍

**Pull request merged**